src/sentry/snuba/metrics/datasource.py

```diff
@@ -158,12 +151,24 @@ def _get_metrics_filter_ids(metric_names: Sequence[str]) -> Set[int]:
     """
     if not metric_names:
         return set()
+
     metric_ids = set()
-    for name in metric_names:
+    metric_names_deque = deque(metric_names)
+    while metric_names_deque:
+        name = metric_names_deque.popleft()
         if name not in DERIVED_METRICS:
             metric_ids.add(indexer.resolve(name))
         else:
-            metric_ids |= DERIVED_METRICS[name].generate_metric_ids()
+            derived_metric_obj = DERIVED_METRICS[name]
+            try:
+                metric_ids |= derived_metric_obj.generate_metric_ids()
+            except NotSupportedOverCompositeEntityException:
+                single_entity_constituents = set(
+                    list(
+                        derived_metric_obj.naively_generate_singular_entity_constituents().values()
+                    ).pop()
+                )
```

Comment:
The set-list-pop is a bit hard to read, what if we hide that in `naively_generate_singular_entity_constituents`? Such that this just becomes 
```suggestion
                single_entity_constituents = derived_metric_obj.naively_generate_singular_entity_constituents()
```

Reply:
Good idea!